src/server/transaction.cc

```diff
@@ -688,7 +688,17 @@ OpStatus Transaction::ScheduleSingleHop(RunnableType cb) {
         CHECK_GE(DecreaseRunCnt(), 1u);
       }
     };
-    shard_set->Add(unique_shard_id_, std::move(schedule_cb));  // serves as a barrier.
+
+    // We can't allow inline scheduling during a full sync, because then journaling transactions
+    // will be scheduled before RdbLoader::LoadItemsBuffer is finished. We can't use the regular
+    // locking mechanism because RdbLoader is not using transactions.
+    if (coordinator_index_ == unique_shard_id_ &&
+        ServerState::tlocal()->gstate() != GlobalState::LOADING) {
```

Comment:
It will be more clear if you check !is_master and sync_in_progress. See the code in ServerFamily::Role

Reply:
I don't think I have a good way to get a reference to the global `Replica` object - it's stored in `ServerFamily` and it's not thread safe to read as well. And if I check `gstate()` then checking `is_master` is a bit redundant, no?